ros2topic/ros2topic/verb/delay.py

```diff
@@ -0,0 +1,236 @@
+#    copyright notice, this list of conditions and the following
+#    disclaimer in the documentation and/or other materials provided
+#    with the distribution.
+#  * Neither the name of Willow Garage, Inc. nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+# This file is originally from:
+# https://github.com/ros/ros_comm/blob/6e5016f4b2266d8a60c9a1e163c4928b8fc7115e/tools/rostopic/src/rostopic/__init__.py
+
+import importlib
+import math
+
+from time import sleep
+
+import rclpy
+
+from rclpy.expand_topic_name import expand_topic_name
+from rclpy.qos import qos_profile_sensor_data
+from rclpy.validate_full_topic_name import validate_full_topic_name
+from ros2cli.node.direct import DirectNode
+from ros2topic.api import get_topic_names_and_types
+from ros2topic.api import TopicNameCompleter
+from ros2topic.verb import VerbExtension
+
+
+class DelayVerb(VerbExtension):
+    """Display delay of topic from timestamp in header."""
+
+    def add_arguments(self, parser, cli_name):
+        arg = parser.add_argument(
+            'topic',
+            help='Topic name to be calcurated the delay')
+        arg.completer = TopicNameCompleter(
+            include_hidden_topics_key='include_hidden_topics')
+        parser.add_argument(
+            '--window', '-w', default=-1, type=int,
+            help='window size, in # of messages, for calculating rate, '
+                 'string to (default: -1)')
+
+    def main(self, *, args):
+        return main(args)
+
+
+def main(args):
+    with DirectNode(args) as node:
+        _rostopic_delay(node.node, args.topic, window_size=args.window, blocking=True)
+
+
+class ROSTopicDelay(object):
+    """Receives messages for a topic and computes timestamp delay."""
+
+    def __init__(self, node, window_size):
+        import threading
+        self.lock = threading.Lock()
+        self.last_msg_tn = 0
+        self.msg_t0 = -1.
+        self.msg_tn = 0
+        self.delays = []
+
+        # can't have infinite window size due to memory restrictions
+        if window_size < 0:
+            window_size = 50000
```

Comment:
This "magic" upper bound is nowhere visible to the user. Please use it as a default value of the argument instead (and ensure in code that the user didn't pass a negative value).

Reply:
this is original design, I have not changed while porting. If user did not set the value. As previous design, the default value of window is -1, and here will re-set to max 50000 as memory restrictions.

Reply:
Sure, it would still benefit from being improved.